Sprint20

vaiwork opened the pull request.

vaiwork committed:
Sprint 20. 2/3 tasks

vaiwork committed:
Sprint 20. 3/3 tasks.

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/data/TracksRepositoryImpl.kt:
```diff
@@ -54,16 +56,16 @@ class TracksRepositoryImpl(
 
             -1 -> {
-                return Resource.Error("Проверьте подключение к интернету")
+                emit(Resource.Error("Проверьте подключение к интернету"))
```
Лучше все-таки избегать хардкодного текста в коде.  В этом приложение оно, конечно, выглядит избыточно, но лучше сразу набивать руку хорошими практиками) Здесь проблему можно решить, добавив дополнительые классы ошибок, то есть не просто error, а конкретно проблема с интернетом. Таким образом не придется в дата слое работать с текстом

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/data/TracksRepositoryImpl.kt:
```diff
@@ -54,16 +56,16 @@ class TracksRepositoryImpl(
-                return Resource.Success(result)
             }
 
             -1 -> {
```
можно вынести в константу, так будет нагляднее, название константы дополнительно подчеркнет смысл

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -8,24 +8,33 @@ import androidx.annotation.RequiresApi
         }
+        return withContext(Dispatchers.IO) {
+            try {
+                //val resp = iTunesService.searchTracks(dto.expression).execute()
```
Если код больше не нужен, то можно удалить совсем

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -8,24 +8,33 @@ import androidx.annotation.RequiresApi
-            return body.apply { resultCode = resp.code() }
-        } else {
+        if (dto !is TrackSearchRequest) {
             return Response().apply { resultCode = 400 } // bad request because request dto unknown
```
Коды ответов тоже можно вынести в константы

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/domain/api/TracksInteractor.kt:
```diff
@@ -1,11 +1,13 @@
 
 interface TracksInteractor {
-    fun searchTracks(expression: String, consumer: TracksConsumer)
+    //fun searchTracks(expression: String, consumer: TracksConsumer)
```
Тоже можно удалить

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/ui/audioplayer/view_model/TracksMediaPlayerViewModel.kt:
```diff
@@ -70,18 +57,32 @@ class TracksMediaPlayerViewModel(
+    private fun startTimer() {
+        timerJob = viewModelScope.launch {
+            while (tracksMediaPlayerInteractor.getPlayerState() == tracksMediaPlayerInteractor.getStatePlaying()) {
+                delay(300L)
```
Лучше вынести длительность в константу и указать единицу времени

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/ui/audioplayer/view_model/TracksMediaPlayerViewModel.kt:
```diff
@@ -70,18 +57,32 @@ class TracksMediaPlayerViewModel(
+                    )
+                )
+            }
+            setSpendTime("00:00")
```
Тоже можно вынести в константу, чем меньше магических чисел и хардкодных строк, тем лучше) Потому что изменять легче в одном месте, чем в нескольких. (А вдруг завтра скажут, что вместо 00:00 надо писать что-то другое?)

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/ui/search/fragment/SearchFragment.kt:
```diff
@@ -32,6 +34,12 @@ class SearchFragment : Fragment() {
 
     private var tracksAdapter: TrackAdapter = TrackAdapter()
 
+    companion object {
```
Согласно конвенции котлина, companion object [должен находиться в конце класса](https://kotlinlang.org/docs/coding-conventions.html#class-layout)

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/ui/search/view_model/TracksSearchViewModel.kt:
```diff
@@ -34,96 +35,81 @@ class TracksSearchViewModel(
+            errorMessage != null -> {
+                renderState(
+                    TracksState.Error(
+                        errorMessage = "Что-то не так",
```
Тоже лучше избавиться от хардкода, лайфхак: можно передавать айдишники из ресурсов, это более валидно

Anna-Malinina commented on app/src/main/java/com/vaiwork/playlistmaker/util/Debounce.kt:
Вот это классное решение, вынести логику)

vaiwork committed:
Sprint 20. After review.
Прогресс бар запускается сразу, а только потом происходит delay - DONE
Стало медленней работать открытие страницы плеера, сильно заметнее - таки специально так сделано, чтобы было видно, что корутина по клику на трек работает (private const val CLICK_DEBOUNCE_DELAY = 10000L)
Если свернуть играющий плеер, а потом открыть заново, то он сбросится - это давнее (все предыдущие спринты) условие работы данного приложения

vaiwork committed:
Sprint 20. After second review.
Возможности, связанные с отправкой поискового запроса, получением результатов поиска, нажатием на элементы списка и отображением прогресса воспроизведения трека работают так же, как и до рефакторинга.
Прогресс бар запускается сразу, а только потом происходит delay. И еще при стирании запроса он тоже сразу включается - исправлено
Треки стали открываться еще медленней (задержка теперь 10 секунд о.о) - теперь 1 секунда
Если свернуть играющий плеер, а потом открыть заново, то он сбросится, а не приостановится. Напомню критерий из 14 спринта: Когда пользователь находится на экране «Аудиоплеер» и происходит воспроизведение отрывка аудиотрека, то при переводе приложения в фоновый режим воспроизведение отрывка аудиотрека приостанавливается, кнопка «Пауза» меняется на «Играть» и обновление текста прогресса воспроизведения останавливается - теперь приостанавливается
